### DISCUSSION TITLE
heap allocation failure, and support custom allocators?

### DISCUSSION THREAD

**Discussion by guidorice** [modular/modular#377] [Q&A] [answered]
What is mojo's philosophy on heap allocation failure, and custom allocators? I was messing around with Pointer.alloc, and could not get it to raise an Error. I did get it to return a zero address, by giving a large enough count parameter. Zig has an interesting take on allocation failures, and also offers a variety of allocators and  allows userland allocators.  https://ziglang.org/documentation/master/#Heap-Allocation-Failure

**Comment by lattner** [comment, marked as answer]
We don't have an established policy here, and I wouldn't read too much into Pointer since it is very much a first cut at something and hasn't been polished.  This is a really complicated topic as I think you know from Zig.  I'm not keen on making everyone /always/ think about allocators like Zig does, I don't think that is practical in a language that cares about usability and ergonomics, but it is clearly good to /allow/ folks to care.

In my personal opinion, there is a big difference practically between "node" allocation and "array" allocation.  Error handling for small objects will kill us, and we don't want to make allocation of any class instance be failable.  That said, allocating an array that could be 16GB definitely can fail.  On the third hand, core data structures like Array probably don't want to expose memory allocation failability to the client by default for usability reasons.

It would be interesting to explore making these different APIs, possibly overloaded with a keyword argument or something.  As one idea, we could make `UnsafePointer[T].allocate()` non-failable, but make `UnsafePointer[T].allocate(Int)` failable.  We'd still have to decide what to do with that at the Array api level, but it too could have overloads for `arr.resize(n)` vs `arr.resize(checked = n)` or something like that.

**Reply by guidorice** [reply]
Love this approach - sounds like a good compromise, and not hurting ergonomics. 👍🏽
